### PR TITLE
Scalars support

### PR DESCRIPTION
**Pull request by YoelShoshan**
added support for scalars as inputs

a 1 -> model_dim projection is learned, used on the inputs, and added (*not* concatenated) on top of the tokens embeddings, only to input elements that have scalar input.

### CODE REVIEW

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -2117,6 +2135,7 @@ def forward(
 
         lm_logits = self.lm_head(sequence_output)
 
+        #import ipdb;ipdb.set_trace()
```
Comment (mosheraboh) [inline src/transformers/models/t5/modeling_t5.py]: remove

File: src/transformers/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -1309,6 +1314,13 @@ def forward(
                 raise ValueError("You have to initialize the model with valid token embeddings")
             inputs_embeds = self.embed_tokens(input_ids)
 
+            if self.project_scalars is not None:
+                for sample_idx, (_curr_scalar_ind, _curr_scalar_vals) in enumerate(zip(encoder_input_scalars_indices, encoder_input_scalars_values)):
+                    assert not ((_curr_scalar_ind is not None) ^ (_curr_scalar_vals is not None)) #must be neither or both
+                    if (_curr_scalar_ind is not None) and (_curr_scalar_ind.shape[0] > 0):
+                        inputs_embeds[sample_idx][_curr_scalar_ind] += self.project_scalars(_curr_scalar_vals[..., None])
```
Comment (mosheraboh) [inline src/transformers/models/t5/modeling_t5.py:1322]: Any option to vectorize it?

File: src/transformers/models/t5/modeling_t5.py [hunk 3]
```diff
@@ -1988,6 +2002,8 @@ def forward(
         return_dict: Optional[bool] = None,
         encoder_position_ids_dict: Optional[Dict[str, Tuple[torch.LongTensor,str]]] = None,  # shape of each LongTensor (num_batches, n_input_tokens)
         decoder_position_ids_dict: Optional[Dict[str, Tuple[torch.LongTensor,str]]] = None, # shape of each LongTensor (num_batches, n_input_tokens)
+        encoder_input_scalars_indices: Optional[torch.LongTensor] = None,
```
Comment (mosheraboh) [inline src/transformers/models/t5/modeling_t5.py]: is it a list of tensors in different sizes or a single padded tensor?

File: src/transformers/models/t5/modeling_t5.py [hunk 4]
```diff
@@ -1309,6 +1314,13 @@ def forward(
                 raise ValueError("You have to initialize the model with valid token embeddings")
             inputs_embeds = self.embed_tokens(input_ids)
 
+            if self.project_scalars is not None:
```
Comment (mosheraboh) [inline src/transformers/models/t5/modeling_t5.py]: You should also check that encoder_input_scalars_indices and encoder_input_scalars_values are not None. 
For a case it was defined but not used by a specific task.

**Review by mosheraboh — APPROVED** [review]
Looks good, see few commnets inline,
Approving this version and per our talk we will work in the future towards moving the logic outside to transformers.

File: src/transformers/models/t5/modeling_t5.py (hunk 2 above)
Comment (YoelShoshan) [reply src/transformers/models/t5/modeling_t5.py:1322]: we can make it vectorized here, but it will be a bit "cheating" because it would mean that as preprocessing it was prepared as a single tensor, and there we had a loop on samples.

there is one possible strategy that we can use, if our collate() was able to also return a mask of valid locations.

either way, not doing this in this PR - let's include this in the PR that moves to calculating embedding outside the transformers and uses inputs_embeds in transformers

File: src/transformers/models/t5/modeling_t5.py (hunk 3 above)
Comment (YoelShoshan) [reply src/transformers/models/t5/modeling_t5.py]: fixed the typing hint

File: src/transformers/models/t5/modeling_t5.py (hunk 4 above)
Comment (YoelShoshan) [reply src/transformers/models/t5/modeling_t5.py]: added, thanks!

File: src/transformers/models/t5/modeling_t5.py (hunk 2 above)
Comment (YoelShoshan) [reply src/transformers/models/t5/modeling_t5.py:1322]: for future reference:
few ways to vectorize it are:

1. Holding it at the size of the full sequence from the start - for large sequences this can have significant overhead (but on the other hand, if it's a large sequence and most of it are outputs scalars then it would be ok)
2. like in the loss calculation, create a collated/padded vector of all of the values and indices.
